File: examples/integrations/webhook-handler.js
/**
 * Webhook Handler Example (JavaScript/Node.js with Express)
 * Demonstrates how to securely verify and handle incoming workflow webhooks
 */

import express from 'express';
import { verifyWebhookSignature } from '@workflow-engine/sdk';

const app = express();
const PORT = process.env.PORT || 3000;
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || 'your_webhook_secret_here';

// IMPORTANT: Webhook signature verification requires the raw request body
// Use express.raw or verify with text body
app.use(express.json({
  verify: (req, res, buf) => {
    // Store raw body for signature verification
    req.rawBody = buf.toString();
  }
}));

// Webhook endpoint
app.post('/webhooks/workflow-events', (req, res) => {
  const signature = req.headers['x-webhook-signature'];

  // Verify HMAC signature
  const isValid = verifyWebhookSignature(
    req.rawBody || JSON.stringify(req.body),
    signature,
    WEBHOOK_SECRET
  );

  if (!isValid) {
    console.error('❌ Invalid webhook signature!');
    return res.status(401).json({ error: 'Invalid signature' });
  }

  // Signature is valid, process the event
  const { event, deliveryId, timestamp, data } = req.body;

  console.log(`✅ Received verified webhook event: ${event}`);
  console.log('Delivery ID:', deliveryId);
  console.log('Timestamp:', timestamp);
  console.log('Execution ID:', data.executionId);
  console.log('Workflow ID:', data.workflowId);
  console.log('Status:', data.status);

  // Handle specific event types
  switch (event) {
    case 'WORKFLOW_EXECUTION_STARTED':
      console.log(`Execution ${data.executionId} has started`);
      break;

    case 'WORKFLOW_EXECUTION_COMPLETED':
      console.log(`Execution ${data.executionId} succeeded!`);
      console.log('Results:', data.result);
      // Perform post-execution actions (e.g., notify user, update database)
      break;

    case 'WORKFLOW_EXECUTION_FAILED':
      console.error(`Execution ${data.executionId} failed!`);
      console.error('Error:', data.error);
      // Perform error recovery actions (e.g., alert team)
      break;

    case 'WORKFLOW_EXECUTION_REPLAYED':
      console.log(`Execution ${data.executionId} is being replayed`);
      break;

    default:
      console.log(`Unhandled event type: ${event}`);
  }

  // Always return 200 OK quickly to acknowledge receipt
  return res.status(200).json({ received: true });
});

app.listen(PORT, () => {
  console.log(`🚀 Webhook server listening on port ${PORT}`);
  console.log(`Endpoint: http://localhost:${PORT}/webhooks/workflow-events`);
});

File: examples/integrations/webhook-handler.py
"""
Webhook Handler Example (Python with Flask)
Demonstrates how to securely verify and handle incoming workflow webhooks
"""

import os
from flask import Flask, request, jsonify
from workflow_engine import verify_webhook_signature

app = Flask(__name__)
WEBHOOK_SECRET = os.getenv("WEBHOOK_SECRET", "your_webhook_secret_here")


@app.route("/webhooks/workflow-events", methods=["POST"])
def handle_webhook():
    # Get the raw request body as string for signature verification
    payload = request.get_data(as_text=True)
    signature = request.headers.get("X-Webhook-Signature")

    # Verify HMAC signature
    is_valid = verify_webhook_signature(payload, signature, WEBHOOK_SECRET)

    if not is_valid:
        print("❌ Invalid webhook signature!")
        return jsonify({"error": "Invalid signature"}), 401

    # Signature is valid, process the event
    data = request.json
    event = data.get("event")
    delivery_id = data.get("deliveryId")
    timestamp = data.get("timestamp")
    payload_data = data.get("data", {})

    print(f"✅ Received verified webhook event: {event}")
    print(f"Delivery ID: {delivery_id}")
    print(f"Timestamp: {timestamp}")
    print(f"Execution ID: {payload_data.get('executionId')}")
    print(f"Workflow ID: {payload_data.get('workflowId')}")
    print(f"Status: {payload_data.get('status')}")

    # Handle specific event types
    if event == "WORKFLOW_EXECUTION_STARTED":
        print(f"Execution {payload_data.get('executionId')} has started")

    elif event == "WORKFLOW_EXECUTION_COMPLETED":
        print(f"Execution {payload_data.get('executionId')} succeeded!")
        print(f"Results: {payload_data.get('result')}")

    elif event == "WORKFLOW_EXECUTION_FAILED":
        print(f"Execution {payload_data.get('executionId')} failed!")
        print(f"Error: {payload_data.get('error')}")

    elif event == "WORKFLOW_EXECUTION_REPLAYED":
        print(f"Execution {payload_data.get('executionId')} is being replayed")

    # Always return 200 OK quickly
    return jsonify({"received": True}), 200


if __name__ == "__main__":
    port = int(os.getenv("PORT", 3000))
    print(f"🚀 Webhook server listening on port {port}")
    app.run(port=port)
